arch-riscv: adding support for uncacheable address ranges in SE mode

iliasslasri opened the pull request:
This PR implements the support for uncacheable ranges in the riscv arch
in SE mode.
In the RISCV MMU, we changed the BasePMAChecker to PMAChecker which
inherents from BasePMAChecker, this class has an attribute
uncacheable_range that we can init in the python script, and implement
functions like ```isUncaheable```that we used to condition the
turning of the flag *UNCHACHEABLE* of the request when translating
the address.

In the mmu, pagetable_walker and tlb, we change the attribute type
BasePMAChcker to PMAChecker (so as in The wrappers RiscvMMU.py and
RiscvTLB.py), then we use this attribute (i.e, pma) to call the function
isUncacheable and condition the setting of the flag
```Request::UNCACHEBLE```.

Change-Id: I68f6ed6f15fda8ca5878e66b581863ec1abcd79a

Signed-off-by: Iliass Lasri <lasri.iliass@huawei.com>
Signed-off-by: Marcelo Ruaro <marcelo.ruaro@huawei.com>

arichardson committed:
arch: Dump semihosting write buffer in debug output (#1389)

This makes it easier to debug unexpected semihosting outputs (in my case
a wrong buffer argument was being passed).

Change-Id: I342610a92fb8efe121d030f7b9ea3307efc4fec3

rogerchang23424 committed:
arch,cpu: Implement generic reset method for MMU (#1342)

Implementing generic reset method for MMU allows each ISA implementing
their own reset methods. The default reset MMU method is flush all TLB
entries. For example, The RISC-V needs to do PMP reset when received the
reset signal, but the TLBs don't require to be flushed.

Change-Id: I158261570fb6e5216ec105fbdc53460f83f88d15

abmerop committed:
misc: Remove GCN3 from maintainers (#1395)

These files no longer exist. This was missed when GCN3 support was
removed in the last release.

Change-Id: I5d2aab1952c37e64da5c362a6201aa6750531c1b

erin-le committed:
sim: Add error message for kernel exceeding memory size (#1329)

This commit adds an error message to src/sim/kernel_workload.cc to tell
the user when the end address of the kernel is greater than the size of
memory. The error message also specifies the minimum memory size needed
to fit the kernel.

Change-Id: I7d8f50889ed8172f64b84f98301a35e5f2f352d3

Commit:
arch-riscv: adding support for uncacheable address ranges in SE mode

This PR implements the support for uncacheable ranges in the riscv arch
in SE mode.
In the RISCV MMU, we changed the BasePMAChecker to PMAChecker which
inherents from BasePMAChecker, this class has an attribute
uncacheable_range that we can init in the python script, and implement
functions like ```isUncaheable```that we used to condition the
turning of the flag *UNCHACHEABLE* of the request when translating
the address.

In the mmu, pagetable_walker and tlb, we change the attribute type
BasePMAChcker to PMAChecker (so as in The wrappers RiscvMMU.py and
RiscvTLB.py), then we use this attribute (i.e, pma) to call the function
isUncacheable and condition the setting of the flag
```Request::UNCACHEBLE```.

Signed-off-by: Iliass Lasri <lasri.iliass@huawei.com>
Signed-off-by: Marcelo Ruaro <marcelo.ruaro@huawei.com>

arichardson committed:
arch-riscv: Fix implicit int-to-float conversion in .isa files (#1319)

Explicitly convert to float/double to fix compiler warnings that I have
turned on locally.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rogerchang23424 commented on src/arch/riscv/tlb.cc:
```diff
@@ -422,6 +428,14 @@ TLB::translate(const RequestPtr &req, ThreadContext *tc,
         if (fault != NoFault)
             return fault;
 
+        // look if the physical address is uncacheable
+        // and set the flag accordingly
+        Addr paddr = req->getPaddr();
+        if(pma->isUncacheable(paddr, req->getSize())){
+            req->setFlags(Request::UNCACHEABLE);
+            DPRINTF(TLB, "Access to address %#x turned to uncacheable.", paddr);
+        }
+
```
I found there some common code section to add the `Request::UNCACHEABLE` flags to Request Packet. 

I think we can have the abstract method `checkSE` in `BasePMAChecker` and implement it in the `PMAChecker::checkSE` like

```cpp
void
PMAChecker::checkSE(const RequestPtr &req)
{
    if (isUncacheable(req->getPaddr(), req->getSize())) {
            req->setFlags(Request::UNCACHEABLE);
            DPRINTF(TLB, "Access to address %#x turned to uncacheable.", req->getPaddr());
    }
}
```

Ref: https://gem5-review.googlesource.com/c/public/gem5/+/40596

It won't require to change the `BasePMAChecker` to `PMAChecker`. How do you think?

rogerchang23424 replied:
@iliasslasri,

Please reference the pervious PR https://github.com/gem5/gem5/pull/846, I feel that this PR revert the BasePMAChecker changes.

ivanaamit commented:
Hi @iliasslasri,

None of the commits in this pull request contains a Change-ID, which we require for any changes made to gem5.  To automatically insert one, run the following:

```
f=.git/hooks/commit-msg
mkdir -p $f
curl -Lo $f https://gerrit-review.googlesource.com/tools/hooks/commit-msg
chmod +x $f
```

Then, amend the commit with `git commit --amend --no-edit`, and update your pull request. Thank you.

hnpl reviewed (commented):
Thanks for the contribution.

I saw that the pma_checker parameter is now of type `PMAChecker` instead of its base class. Is there a particular reason for this change?

iliasslasri commented:
Hi @hnpl,
Indeed, I changed the type of the pma_checker from ```BasePMAChecker``` to ```PMAChecker```, to use the extra attribute ```AddrRangeList uncacheable;```, to add the support of uncacheable ranges without adding attributes or function to any class. The PMAChecker inherits from BasePMAChecker, so there will be no compatibility issue.

dependabot[bot] committed:
misc: bump pre-commit from 3.7.1 to 3.8.0 (#1408)

Bumps [pre-commit](https://github.com/pre-commit/pre-commit) from 3.7.1
to 3.8.0.


Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

giactra committed:
mem: Fix name() helper for DRAM rank (#1410)

At the moment the method simply returns the rank number. This is not
particularly useful when enabling debug flags as the beginning of the
line prints something like:

1: <debug_message>

whereas it should really be:

system.dram.rank1: <debug_message>

Change-Id: I0136dc3d182afa4ae2e5a719cb366d8d0f444667

Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

iliasslasri committed:
Merge branch 'develop' into rv-uncacheable-ranges

powerjg commented on src/arch/riscv/RiscvMMU.py:
```diff
@@ -49,7 +49,7 @@ class RiscvMMU(BaseMMU):
     itb = RiscvTLB(entry_type="instruction")
     dtb = RiscvTLB(entry_type="data")
-    pma_checker = Param.BasePMAChecker(PMAChecker(), "PMA Checker")
+    pma_checker = Param.PMAChecker("PMA Checker")
```
I strongly agree with this change to remove an instance of a SimObject as the default value for a parameter.

However, I worry that there are configuration scripts that depend on this. We need to test the stdlib boards and the config scripts in `configs/examples` before we can merge this change.

In fact, it looks like this is likely why the CI tests are failing. See https://github.com/gem5/gem5/actions/runs/10286529082/job/28476297182?pr=1401

ivanaamit commented:
Hi @iliasslasri, you need to fix the failing tests in this PR to get it closer to being merged. Thanks.

ivanaamit commented:
Hi @iliasslasri, please let me know if you're still working on this. If not, I'll close the PR, but you can reopen it anytime if you decide to continue. Thanks!

erin-le commented:
Closing due to inactivity. This PR can be reopened if someone wants to work on it in the future.
